### PR TITLE
Write lines to their original levels so that /window hidelevel can be…

### PR DESCRIPTION
**Pull request by vague666**
… used on the pre-upgrade buffer

### CODE REVIEW

**Comment by ailin-nemui** [comment]
we need to change this script in such a way that it still works fine on old irssis...

**Comment by vague666** [comment]
I haven't tested the script without #862 applied, I don't know how critical the patch is

**Comment by ailin-nemui** [comment]
forget what I said but please don't change indentation (4 spaces) of existing scripts ... :-) I will add more review comments once you can do that

File: scripts/buf.pl [hunk 1]
```diff
@@ -40,64 +43,59 @@
 sub _filename { sprintf '%s/scrollbuffer', get_irssi_dir }
 
 sub upgrade {
-    my $fn = _filename;
-    my $old_umask = umask 0077;
-    open my $fh, q{>}, $fn or die "open $fn: $!";
-    umask $old_umask;
-
-    print $fh join("\0", map $_->{server}->{address} . $_->{name}, channels), "\n";
+    my $out = { suppress => join("\0", map $_->{server}->{address} . $_->{name}, channels) };
```
Comment (ailin-nemui) [inline scripts/buf.pl]: how about you replace this with a structure / array instead of the awkward NUL byte

**Comment by vague666** [comment]
That can be done yeah, didn't even think of it, tried to keep as much of the old script as possible but yeah, it makessense now that we use Storable

File: scripts/buf.pl [hunk 2]
```diff
@@ -40,64 +43,59 @@
...
-
-    my @suppress = split /\0/, readline $fh;
-    if (settings_get_bool 'upgrade_suppress_join') {
-	chomp $suppress[-1];
-	@suppress{@suppress} = (2) x @suppress;
-    }
+    my $in = retrieve($fn) or die $!;
+    unlink $fn;
+  
+    my @suppress = split /\0/, $in->{suppress};
```
Comment (ailin-nemui) [inline scripts/buf.pl]: ... then you would need to remove this split here

File: scripts/buf.pl [hunk 3]
```diff
@@ -40,64 +43,59 @@
...
+              push @$output, { level => $line->{info}{level}, data => $line->get_text(1) };
+              $line = $line->next;
+              redo if defined $line;
+            }
+        }
+        push @{$out->{windows}}, { refnum => $window->{refnum}, lines => $output };
     }
-    close $fh;
+    my $old_umask = umask 0077;
+    store($out, _filename) or die $!;
```
Comment (ailin-nemui) [inline scripts/buf.pl]: I can't find any mention in the docs that Storable sets $!. I think you would have to die with a generic error string ("storing failed")

File: scripts/buf.pl [hunk 4]
```diff
@@ -40,64 +43,59 @@
...
+        next unless $window;
+        my @lines  = @{ $win->{lines} || [] };
+        next unless @lines;
+    
+        my $view = $window->view;
+        $view->remove_all_lines();
+        $view->redraw();
+        my $sep = settings_get_str 'upgrade_separator';
+        $sep .= "\n" if $sep ne '';
+        $window->gui_printtext_after(undef, MSGLEVEL_CLIENTNOTICE, "\cO$sep\n");
```
Comment (ailin-nemui) [inline scripts/buf.pl]: this separator notice should go LAST after all the other lines have been printed in each window, at least that's how it originally was

File: scripts/buf.pl [hunk 5]
```diff
@@ -40,64 +43,59 @@
...
-    open my $fh, q{<}, $fn or die "open $fn: $!";
-    unlink $fn or warn "unlink $fn: $!";
-
-    my @suppress = split /\0/, readline $fh;
-    if (settings_get_bool 'upgrade_suppress_join') {
-	chomp $suppress[-1];
-	@suppress{@suppress} = (2) x @suppress;
-    }
+    my $in = retrieve($fn) or die $!;
+    unlink $fn;
```
Comment (ailin-nemui) [inline scripts/buf.pl]: why did you remove the warning here??

File: scripts/buf.pl (hunk 4 above)
Comment (vague666) [reply scripts/buf.pl]: iirc the screen got scrolled up if I did it that way so the =Upgrade= wasn't visible

Comment (ailin-nemui) [reply scripts/buf.pl]: move it below the loop and use $window->last_line_insert instead of undef, should work I think

File: scripts/buf.pl (hunk 5 above)
Comment (vague666) [reply scripts/buf.pl]: It must've gone out with speed, no solid reason as to why, I readded it :)

File: scripts/buf.pl [hunk 6]
```diff
@@ -40,64 +42,60 @@
...
+        my $view = $window->view;
+        $view->remove_all_lines();
+        $view->redraw();
+        for my $line (@lines) {
+            my $level = $line->{level};
+            my $data  = $line->{data};
+            $window->gui_printtext_after($window->last_line_insert, $level, "$data\n");
+        }
+        my $sep = settings_get_str 'upgrade_separator';
+        $sep .= "\n" if $sep ne '';
```
Comment (ailin-nemui) [inline scripts/buf.pl]: this line seems to cause an empty line to appear on screen

Comment (ailin-nemui) [reply scripts/buf.pl]: maybe just remove it?

Comment (vague666) [reply scripts/buf.pl]: Maybe we should just remove it?

Comment (vague666) [reply scripts/buf.pl]: Or remove the \n from 
$window->gui_printtext_after($window->last_line_insert, MSGLEVEL_CLIENTNOTICE, "\cO$sep\n");

Comment (ailin-nemui) [reply scripts/buf.pl]: that seems buggy (bug was in the original script) if we have an empty $sep then we should not print anything (also not a \c0) otherwise we have an incomplete line hanging around in the buffer

Comment (vague666) [reply scripts/buf.pl]: Alright, that should be it :)

**Comment by ailin-nemui** [comment]
you need to test "if $sep", defined is too strong. it will be defined but empty string

**Comment by vague666** [comment]
That should do it, I used the defined to avoid $sep being '0', but didn't think of the consequences

**Comment by ailin-nemui** [comment]
ok shall we merge it then?

**Comment by vague666** [comment]
Maybe run a quick test first? It looked like you tested it as we went along here in the end

**Comment by ailin-nemui** [comment]
please test it and tell me if everything is fine ^^

**Comment by vague666** [comment]
And we are still using gui_printtext_after, should we change it to straight prnt instead? Do we have access to $Irssi::VERSION or whatever?

**Comment by vague666** [comment]
Appears to work, it gets my thumbs up
